src/linker/Linker.Steps/MarkStep.cs

```diff
@@ -2928,6 +2928,9 @@ protected virtual void MarkInstruction (Instruction instruction, MethodDefinitio
 					if (type.IsInterface)
 						break;
 
+					if (!_context.IsOptimizationEnabled (CodeOptimizations.UnusedTypeChecks, method.DeclaringType))
+						break;
+
 					if (!Annotations.IsInstantiated (type)) {
```

Comment:
Nit: I would merge the two conditions - to make it obvious that it's about the same thing. If we were to add additional processing here, it would be easy to make the mistake of leaving the optimization check as is and hide another functionality behind it.
But it's a very minor thing.